app/services/patient_deduplication/deduplicator.rb

```diff
@@ -0,0 +1,212 @@
+module PatientDeduplication
+  class Deduplicator
+    def initialize(patients, user: nil)
+      @patients = patients.sort_by(&:recorded_at)
+      @user_id = user&.id
+    end
+
+    def earliest_patient
+      @patients.first
+    end
+
+    def latest_patient
+      @patients.last
+    end
+
+    def merge
+      patient_ids = @patients.pluck(:id)
+      Rails.logger.info "Merging patients #{patient_ids}"
+
+      new_patient = ActiveRecord::Base.transaction do
+        new_patient = create_patient
+        create_prescription_drugs(new_patient)
+        create_medical_history(new_patient)
+        create_phone_numbers(new_patient)
+        create_patient_business_identifiers(new_patient)
+        create_encounters_and_observables(new_patient)
+        create_appointments(new_patient)
+        create_teleconsultations(new_patient)
+        mark_as_merged(new_patient)
+        new_patient.reload
+      end
+
+      Rails.logger.info "Merged patients #{patient_ids} into patient #{new_patient.id}"
+      new_patient
+    end
+
+    def mark_as_merged(new_patient)
+      @patients.map do |patient|
+        patient.update!(
+          merged_into_patient_id: new_patient.id,
+          merged_by_user_id: @user_id
+        )
+      end
+      @patients.map(&:discard_data)
```

Comment:
Our discard_data calls only `discard` methods, not `discard`.  That means we could do all the merges, and then attempt to discard, which will fail silently due to validation issues. We could easily end up in a situation where we have the new merged records, and some of the old invalid non-discarded records still hanging out.

I think we need to do a `discard!` for all the old dupe records to make sure we don't end up in a really bad state here.  It looks like [discard](https://github.com/jhawthorn/discard) supports the usual AR bang semantics for this.

Reply:
The `update!` above should take care of any failing `patient` validations. For associated models, the only change we make is to update the `deleted_at`. This should be okay? (unless the data was already bad).

I think we should separately make `discard_data` use bang methods if we want to fail since it's ubiquitously used. wdyt?

Reply:
Created a [card on clubhouse](https://app.clubhouse.io/simpledotorg/story/2958/handle-invalid-records-in-patient-discard-data) to handle this later.